### DISCUSSION TITLE
About Secret OTP

### DISCUSSION THREAD

**Discussion by andika-bfi** [keycloakify/keycloakify#299] [General]
I am using this library in version 5.7.3, and after building, there is a secretotp key when inspected in the browser, for example, in Chrome, when going to the source tab, there is:

`totpSecretEncoded: "KVVF G2BY N4YX S6LB IUYT K2LH IFYE 4SBV"`

I would like to ask, what is the purpose of the secret OTP? Is it dangerous in case of theft?

Thank you.

**Comment by garronej** [comment]
@andika-bfi,  

When there is `secret` in the name of the variable, the rule of thumbs is it shouldn't be public.  
At the same time, there shouldn't be any sensible information made available in the FTL context and I assume the Keycloak team knows what they are doing to that regard.  
Anyway I can hide that at a moment’s notice.  

@remigauthierdocaposte @pedrodsRocha1 What do you think, should I hide this?

**Reply by remigauthierdocaposte** [reply]
In this case the totpSecretEncoded is the secret key embeded in the QRCode to register the totp.
It will be displayed in plain text in the frontend in case the QRCode cannot be used and the used asks to display it.
IMO it should stay visible.

**Reply by remigauthierdocaposte** [reply]
To be more precise this "totpSecretEncoded" is only present here : 
https://github.com/keycloakify/keycloakify/search?q=totpSecretEncoded

Excepted the mock page, it is only used on the screen to configure totp : `login-config-totp.ftl`
This is the expected behavior and the "secret" value is meant to be displayed for (and scanned by) the user to register his TOTP.

**Reply by garronej** [reply]
Great thank you very much @remigauthierdocaposte for taking the time to investigate this. You're of great help.

**Reply by andika-bfi** [reply]
Is there any suggestion to prevent mock code/ "totpsecret" file from being compiled in React's build files?  Because I'm not using an OTP template.

thank you

**Reply by garronej** [reply]
I'm not sure I understand what you mean

**Reply by joskoanicic** [reply]
@garronej  sorry for reviving a zombie thread, but as we have encountered this issue during our pen testing I would like to add to this matter and try to clarify what andika-bfi was probably asking.

In the main.js file there is mock data being compiled (eg. below), not just for the login-config-totp.ftl but for all other ftl templates. 
```javascript
 z.id)(Object.assign(Object.assign({}, R), {
                pageId: "login-idp-link-email.ftl",
                idpAlias: "FranceConnect",
                brokerContext: {
                    username: "anUsername"
                }
            })), (0,
            z.id)(Object.assign(Object.assign({}, R), {
                pageId: "login-config-totp.ftl",
                totp: {
                    totpSecretEncoded: "KVVF G2BY N4YX S6LB IUYT K2LH IFYE 4SBV",
                    qrUrl: "#",
                    totpSecretQrCode: "iVBORw0KG....1BYMAAAAASUVORK5CYII=",
                    manualUrl: "#",
                    totpSecret: "G4nsI8lQagRMUchH8jEG",
                    otpCredentials: [],
                    supportedApplications: ["FreeOTP", "Google Authenticator"],
                    policy: {
                        algorithm: "HmacSHA1",
                        digits: 6,
                        lookAheadWindow: 1,
                        type: "totp",
                        period: 30
                    }
                }
            })), 
```
Is this necessary or can it be prevented to show up in the main.js file?  This shows up in penetration testing as secret data leaking, although it is not as this is mock data but it is hard to explain to a lot of different security teams out there. By removing all the mock data we get the benefit of having a more lean main.js file generated. 

Thank you.

**Reply by garronej** [reply]
Hello @joskoanicic,  

This is a good point. And an easy to address for one.  

Infact the mock data provided in the KcContext.ts are meerly a convignence for having a baseline for the mock context.  

https://github.com/keycloakify/keycloakify-starter/blob/04e19af690427faca26c4946a491bbc2bf945f81/src/keycloak-theme/login/kcContext.ts#L18-L90  

In reality the best place to define mocks are in the Storybook stories files, like for example here:  

https://github.com/keycloakify/keycloakify-starter/blob/04e19af690427faca26c4946a491bbc2bf945f81/src/keycloak-theme/login/pages/Login.stories.tsx#L62-L80   

So I sugest moving your mock data to the stories. You can probably remove the `"mockData": [` property from the `kcContext.ts` file.

**Reply by joskoanicic** [reply]
Hi @garronej,

by removing mockData from kcContext.ts removes the mock data object from the beginning of the main.js file but the many Object.assign calls with mock data still remain.

I traced the source of mock data to be from:
https://github.com/keycloakify/keycloakify/blob/0796b3dedfc4f1799a33621375ec9d7178fcb5a9/src/login/kcContext/kcContextMocks.ts#L590-L610

Can these be disabled somehow? 
Thank you.

**Reply by garronej** [reply]
Ok I see the issue, I'll see what I can do. It will be a breaking change however.

**Reply by garronej** [reply]
@joskoanicic,  
I've implemented a big refactor that prevent any mock data to be included in the dist.  
It will be available in Keycloakify 10
Thanks for the heads up. 

Ref: https://github.com/keycloakify/keycloakify/pull/538/commits/c1a63edd71f6337df73a83966a8f283e7f2d8c7b

**Reply by joskoanicic** [reply]
Thank you for the quick response.

**Reply by raghavakrao** [reply]
Hi, using 7 version and our VAPT team reporting about the mock data present in kcContextMock.ts file. Shall I upgrade to 10version to get rid of this issue or any alternate? worry is I may trap into more vulnerabilities or conflicts with existing flow.
